lighthouse-core/gather/gatherers/js-usage.js

```diff
@@ -28,33 +35,48 @@ class JsUsage extends Gatherer {
   }
 
   /**
-   * @param {LH.Gatherer.PassContext} passContext
+   * @param {LH.Gatherer.FRTransitionalContext} context
    */
-  async beforePass(passContext) {
-    await passContext.driver.sendCommand('Profiler.enable');
-    await passContext.driver.sendCommand('Profiler.startPreciseCoverage', {detailed: false});
-
-    await passContext.driver.sendCommand('Debugger.enable');
-    await passContext.driver.on('Debugger.scriptParsed', this.onScriptParsed);
+  async startInstrumentation(context) {
+    const session = context.driver.defaultSession;
+    await session.sendCommand('Profiler.enable');
+    await session.sendCommand('Profiler.startPreciseCoverage', {detailed: false});
   }
 
   /**
-   * @param {LH.Gatherer.PassContext} passContext
-   * @return {Promise<LH.Artifacts['JsUsage']>}
+   * @param {LH.Gatherer.FRTransitionalContext} context
    */
-  async afterPass(passContext) {
-    const driver = passContext.driver;
+  async stopInstrumentation(context) {
+    const session = context.driver.defaultSession;
+    const coverageResponse = await session.sendCommand('Profiler.takePreciseCoverage');
+    this._scriptUsages = coverageResponse.result;
+    await session.sendCommand('Profiler.stopPreciseCoverage');
+    await session.sendCommand('Profiler.disable');
+  }
 
-    const coverageResponse = await driver.sendCommand('Profiler.takePreciseCoverage');
-    const scriptUsages = coverageResponse.result;
-    await driver.sendCommand('Profiler.stopPreciseCoverage');
-    await driver.sendCommand('Profiler.disable');
+  /**
+   * @param {LH.Gatherer.FRTransitionalContext} context
+   */
+  async startSensitiveInstrumentation(context) {
+    const session = context.driver.defaultSession;
+    await session.sendCommand('Debugger.enable');
+    await session.on('Debugger.scriptParsed', this.onScriptParsed);
+  }
 
-    await passContext.driver.sendCommand('Debugger.disable');
+  /**
+   * @param {LH.Gatherer.FRTransitionalContext} context
+   */
+  async stopSensitiveInstrumentation(context) {
+    await context.driver.defaultSession.sendCommand('Debugger.disable');
+  }
 
+  /**
+   * @return {Promise<LH.Artifacts['JsUsage']>}
```

Comment:
this artifact is also used by snapshots just to lookup the scriptId -> URL mapping, any thoughts on splitting the artifact or conditionally enabling the profiler to achieve those results?

don't need to do anything extra in this PR, but we should annotate it somehow that there's room for upgrading here (FR-COMPAT or in the inventory sheet in a "notes" column?)

Reply:
Don't we get our scriptId -> URL mappings from the debugger not the profiler? I don't see how we can use the debugger in snapshot mode since we wait for `ScriptParsed` events. The profiler can give us similar results but will be subject to interference by `scriptURL` comments.

Reply:
Correct it comes from the debugger. With most domains though, you can run `Debugger.enable` after a page has loaded and still receive many of the relevant events.

For example...

```js
const puppeteer = require('puppeteer')

;(async () => {

const browser = await puppeteer.launch()
const page = await browser.newPage()

await page.goto('https://www.optyx.app', {waitUntil: ['networkidle0']})
await new Promise(r => setTimeout(r, 5000))

const session = await page.target().createCDPSession()
session.on('Debugger.scriptParsed', e => console.log(e.url))
await session.send('Debugger.enable')

await browser.close()
})()
```

will still log all of the page's javascript

Reply:
(that's partly how DevTools is still able to work when you open it after page load ;)

Reply:
Ah ok. How can we guarantee all the `ScriptParsed` events are handled by the time we use them in `getArtifact`?

Reply:
> With most domains though

What determines if the events are emitted or not?

Reply:
> How can we guarantee all the ScriptParsed events are handled by the time we use them in getArtifact?

Good question. 

1. Investigate if the domain enable command waits for the historical events to be emitted before returning (I don't know if this is the case, but it'd be cool if it were and I vaguely remember some discussion about this a number of years ago with DevTools. It's possible the discussion was "Patrick: Hey what if it worked like *this*? DevTools: lulz, no way", but it's possible 😅 
2. Repurpose a "network quiet" style wait period with a timeout.

Either way I don't think you need to solve it in this PR as it's an upgrade, but these are the sort of things we should be noting as we go, so we have a nice handy burndown ready to go when we make our "upgrade" pass over the gatherers :)

Reply:
> What determines if the events are emitted or not?

I think it's based on whether Chromium had to keep this data around in memory anyway and/or DevTools making special exceptions for the important open DevTools on existing page use case and still get useful information. I also remember discussing at some point the idea that there'd be a lot more DevTools could do this with if it had a "Developer mode" signal or something, but I'm pretty disconnected from all of that at this point, so I'm not sure what's going on these days in this area.

Reply:
Ok I'll make the notes in the conversion doc and add a comment here. I think adding a special case for snapshot would be preferable to splitting this gatherer.